File: app/src/main/kotlin/net/tigr/soulcalc/ui/component/CalculatorKeyboard.kt
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 Albert Zenkoff
 */

package net.tigr.soulcalc.ui.component

import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import net.tigr.soulcalc.R

/**
 * Custom calculator keyboard.
 *
 * Layout:
 * ```
 *  ^    #   $   =   ⌫
 *  7    8   9   ÷   (
 *  4    5   6   ×   )
 *  1    2   3   −   ⏎ (tall)
 *  0    .   %   +   ⏎ (tall)
 * ```
 *
 * Long-press ^ for sqrt(
 * Long-press ⌫ to clear line
 * Long-press $ to insert result reference
 */
@Composable
fun CalculatorKeyboard(
    onKeyPress: (String) -> Unit,
    onEnter: () -> Unit,
    onBackspace: () -> Unit,
    onClearLine: () -> Unit,
    onDollarKey: () -> Unit,
    onDollarLongPress: () -> Unit,
    onHashKey: () -> Unit,
    modifier: Modifier = Modifier
) {
    val topRows = listOf(
        listOf(
            KeyDef("^", KeyType.POWER),
            KeyDef("#", KeyType.SPECIAL_HASH),
            KeyDef("$", KeyType.SPECIAL_DOLLAR),
            KeyDef("=", KeyType.OPERATOR),
            KeyDef("⌫", KeyType.BACKSPACE)
        ),
        listOf(
            KeyDef("7", KeyType.DIGIT),
            KeyDef("8", KeyType.DIGIT),
            KeyDef("9", KeyType.DIGIT),
            KeyDef("÷", KeyType.OPERATOR),
            KeyDef("(", KeyType.OPERATOR)
        ),
        listOf(
            KeyDef("4", KeyType.DIGIT),
            KeyDef("5", KeyType.DIGIT),
            KeyDef("6", KeyType.DIGIT),
            KeyDef("×", KeyType.OPERATOR),
            KeyDef(")", KeyType.OPERATOR)
        )
    )

    val bottomRows = listOf(
        listOf(
            KeyDef("1", KeyType.DIGIT),
            KeyDef("2", KeyType.DIGIT),
            KeyDef("3", KeyType.DIGIT),
            KeyDef("−", KeyType.OPERATOR)
        ),
        listOf(
            KeyDef("0", KeyType.DIGIT),
            KeyDef(".", KeyType.DIGIT),
            KeyDef("%", KeyType.OPERATOR),
            KeyDef("+", KeyType.OPERATOR)
        )
    )

    Surface(
        modifier = modifier.fillMaxWidth(),
        color = MaterialTheme.colorScheme.surfaceVariant,
        tonalElevation = 2.dp
    ) {
        Column(
            modifier = Modifier.padding(4.dp),
            verticalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            // Top 3 rows (5 buttons each)
            topRows.forEach { row ->
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(4.dp)
                ) {
                    row.forEach { keyDef ->
                        KeyButton(
                            keyDef = keyDef,
                            onTap = {
                                when (keyDef.type) {
                                    KeyType.DIGIT, KeyType.OPERATOR -> onKeyPress(keyDef.label)
                                    KeyType.POWER -> onKeyPress("^")
                                    KeyType.BACKSPACE -> onBackspace()
                                    KeyType.ENTER -> onEnter()
                                    KeyType.SPECIAL_DOLLAR -> onDollarKey()
                                    KeyType.SPECIAL_HASH -> onHashKey()
                                }
                            },
                            onLongPress = when (keyDef.type) {
                                KeyType.BACKSPACE -> onClearLine
                                KeyType.SPECIAL_DOLLAR -> onDollarLongPress
                                KeyType.POWER -> { { onKeyPress("sqrt(") } }
                                else -> null
                            },
                            modifier = Modifier.weight(1f)
                        )
                    }
                }
            }

            // Bottom 2 rows with tall Enter button
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(4.dp)
            ) {
                // Left side: 4 buttons per row, stacked
                Column(
                    modifier = Modifier.weight(4f),
                    verticalArrangement = Arrangement.spacedBy(4.dp)
                ) {
                    bottomRows.forEach { row ->
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.spacedBy(4.dp)
                        ) {
                            row.forEach { keyDef ->
                                KeyButton(
                                    keyDef = keyDef,
                                    onTap = { onKeyPress(keyDef.label) },
                                    onLongPress = null,
                                    modifier = Modifier.weight(1f)
                                )
                            }
                        }
                    }
                }

                // Right side: tall Enter button
                KeyButton(
                    keyDef = KeyDef("⏎", KeyType.ENTER),
                    onTap = onEnter,
                    onLongPress = null,
                    modifier = Modifier
                        .weight(1f)
                        .height(116.dp)  // 2 * 56dp + 4dp gap
                )
            }
        }
    }
}

private enum class KeyType {
    DIGIT,
    OPERATOR,
    POWER,
    BACKSPACE,
    ENTER,
    SPECIAL_DOLLAR,
    SPECIAL_HASH
}

private data class KeyDef(
    val label: String,
    val type: KeyType
)

/**
 * Returns the accessibility content description for a key.
 */
@Composable
private fun keyContentDescription(keyDef: KeyDef): String {
    return when (keyDef.label) {
        "0", "1", "2", "3", "4", "5", "6", "7", "8", "9" ->
            stringResource(R.string.key_digit, keyDef.label)
        "+" -> stringResource(R.string.key_add)
        "−" -> stringResource(R.string.key_subtract)
        "×" -> stringResource(R.string.key_multiply)
        "÷" -> stringResource(R.string.key_divide)
        "^" -> stringResource(R.string.key_power_sqrt)
        "%" -> stringResource(R.string.key_percent)
        "=" -> stringResource(R.string.key_equals)
        "(" -> stringResource(R.string.key_open_paren)
        ")" -> stringResource(R.string.key_close_paren)
        "." -> stringResource(R.string.key_decimal)
        "⌫" -> stringResource(R.string.key_backspace)
        "⏎" -> stringResource(R.string.key_enter)
        "$" -> stringResource(R.string.key_dollar)
        "#" -> stringResource(R.string.key_hash)
        else -> keyDef.label
    }
}

private const val LONG_PRESS_THRESHOLD_MS = 400L

@Composable
private fun KeyButton(
    keyDef: KeyDef,
    onTap: () -> Unit,
    onLongPress: (() -> Unit)?,
    modifier: Modifier = Modifier
) {
    val haptic = LocalHapticFeedback.current
    val keyDescription = keyContentDescription(keyDef)
    var showAltLabel by remember { mutableStateOf(false) }

    val backgroundColor = when (keyDef.type) {
        KeyType.DIGIT -> MaterialTheme.colorScheme.surface
        KeyType.OPERATOR, KeyType.POWER -> MaterialTheme.colorScheme.secondaryContainer
        KeyType.BACKSPACE, KeyType.ENTER -> MaterialTheme.colorScheme.errorContainer
        KeyType.SPECIAL_DOLLAR, KeyType.SPECIAL_HASH -> MaterialTheme.colorScheme.secondaryContainer
    }

    val textColor = when (keyDef.type) {
        KeyType.DIGIT -> MaterialTheme.colorScheme.onSurface
        KeyType.OPERATOR, KeyType.POWER -> MaterialTheme.colorScheme.onSecondaryContainer
        KeyType.BACKSPACE, KeyType.ENTER -> MaterialTheme.colorScheme.onErrorContainer
        KeyType.SPECIAL_DOLLAR, KeyType.SPECIAL_HASH -> MaterialTheme.colorScheme.onSecondaryContainer
    }

    val displayLabel = if (showAltLabel && keyDef.type == KeyType.POWER) "√" else keyDef.label

    Surface(
        modifier = modifier
            .height(56.dp)
            .testTag("key_${keyDef.label}")
            .semantics { contentDescription = keyDescription }
            .pointerInput(keyDef, onTap, onLongPress) {
                detectTapGestures(
                    onPress = { _ ->
                        if (onLongPress != null) {
                            coroutineScope {
                                val longPressJob = launch {
                                    delay(LONG_PRESS_THRESHOLD_MS)
                                    showAltLabel = true
                                    haptic.performHapticFeedback(HapticFeedbackType.LongPress)
                                }
                                val released = tryAwaitRelease()
                                longPressJob.cancel()
                                if (released && showAltLabel) {
                                    onLongPress()
                                } else if (released) {
                                    haptic.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                                    onTap()
                                }
                                showAltLabel = false
                            }
                        } else {
                            val released = tryAwaitRelease()
                            if (released) {
                                haptic.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                                onTap()
                            }
                        }
                    }
                )
            },
        color = backgroundColor,
        shape = MaterialTheme.shapes.small
    ) {
        Box(
            contentAlignment = Alignment.Center
        ) {
            Text(
                text = displayLabel,
                color = textColor,
                fontSize = 20.sp,
                fontWeight = FontWeight.Medium
            )
        }
    }
}

File: ios/SoulCalc/UI/Component/CalculatorKeyboard.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Albert Zenkoff

import SwiftUI
#if canImport(UIKit)
import UIKit
#endif

/// Custom calculator keyboard.
///
/// Layout:
/// ```
///  ^    #   $   =   ⌫
///  7    8   9   ÷   (
///  4    5   6   ×   )
///  1    2   3   −   ⏎ (tall)
///  %    0   .   +   ⏎ (tall)
/// ```
///
/// Long-press ^ for sqrt(
/// Long-press ⌫ to clear line
/// Long-press $ to show variable picker
struct CalculatorKeyboard: View {
    let onKeyPress: (String) -> Void
    let onEnter: () -> Void
    let onBackspace: () -> Void
    let onClearLine: () -> Void
    let onDollarKey: () -> Void
    let onDollarLongPress: () -> Void
    let onHashKey: () -> Void

    private let topRows: [[KeyDef]] = [
        [
            KeyDef("^", .power),
            KeyDef("#", .specialHash),
            KeyDef("$", .specialDollar),
            KeyDef("=", .operator_),
            KeyDef("⌫", .backspace)
        ],
        [
            KeyDef("7", .digit),
            KeyDef("8", .digit),
            KeyDef("9", .digit),
            KeyDef("÷", .operator_),
            KeyDef("(", .operator_)
        ],
        [
            KeyDef("4", .digit),
            KeyDef("5", .digit),
            KeyDef("6", .digit),
            KeyDef("×", .operator_),
            KeyDef(")", .operator_)
        ]
    ]

    private let bottomRows: [[KeyDef]] = [
        [
            KeyDef("1", .digit),
            KeyDef("2", .digit),
            KeyDef("3", .digit),
            KeyDef("−", .operator_)
        ],
        [
            KeyDef("%", .operator_),
            KeyDef("0", .digit),
            KeyDef(".", .digit),
            KeyDef("+", .operator_)
        ]
    ]

    private let spacing: CGFloat = 4
    private let buttonHeight: CGFloat = 56
    private let columns = 5

    var body: some View {
        GeometryReader { geometry in
            let totalSpacing = spacing * CGFloat(columns - 1)
            let buttonWidth = (geometry.size.width - totalSpacing) / CGFloat(columns)

            VStack(spacing: spacing) {
                // Top 3 rows (5 buttons each)
                ForEach(0..<topRows.count, id: \.self) { rowIdx in
                    HStack(spacing: spacing) {
                        ForEach(topRows[rowIdx], id: \.label) { keyDef in
                            KeyButton(
                                keyDef: keyDef,
                                onTap: { handleTap(keyDef) },
                                onLongPress: longPressAction(for: keyDef),
                                width: buttonWidth,
                                height: buttonHeight
                            )
                        }
                    }
                }

                // Bottom 2 rows with tall Enter button
                HStack(spacing: spacing) {
                    // Left side: 4 buttons per row
                    VStack(spacing: spacing) {
                        ForEach(0..<bottomRows.count, id: \.self) { rowIdx in
                            HStack(spacing: spacing) {
                                ForEach(bottomRows[rowIdx], id: \.label) { keyDef in
                                    KeyButton(
                                        keyDef: keyDef,
                                        onTap: { onKeyPress(keyDef.label) },
                                        onLongPress: nil,
                                        width: buttonWidth,
                                        height: buttonHeight
                                    )
                                }
                            }
                        }
                    }

                    // Tall Enter button
                    KeyButton(
                        keyDef: KeyDef("⏎", .enter),
                        onTap: onEnter,
                        onLongPress: nil,
                        width: buttonWidth,
                        height: buttonHeight * 2 + spacing
                    )
                }
            }
        }
        .frame(height: buttonHeight * 5 + spacing * 4)
        .padding(4)
        #if canImport(UIKit)
        .background(Color(.systemGray5))
        #else
        .background(Color(NSColor.windowBackgroundColor))
        #endif
    }

    private func handleTap(_ keyDef: KeyDef) {
        switch keyDef.type {
        case .digit, .operator_:
            onKeyPress(keyDef.label)
        case .power:
            onKeyPress("^")
        case .backspace:
            onBackspace()
        case .enter:
            onEnter()
        case .specialDollar:
            onDollarKey()
        case .specialHash:
            onHashKey()
        }
    }

    private func longPressAction(for keyDef: KeyDef) -> (() -> Void)? {
        switch keyDef.type {
        case .backspace:
            return onClearLine
        case .specialDollar:
            return onDollarLongPress
        case .power:
            return { onKeyPress("sqrt(") }
        default:
            return nil
        }
    }
}

// MARK: - Key Types

private enum KeyType {
    case digit
    case operator_
    case power
    case backspace
    case enter
    case specialDollar
    case specialHash
}

private struct KeyDef {
    let label: String
    let type: KeyType

    init(_ label: String, _ type: KeyType) {
        self.label = label
        self.type = type
    }
}

// MARK: - Key Button

private struct KeyButton: View {
    let keyDef: KeyDef
    let onTap: () -> Void
    let onLongPress: (() -> Void)?
    var width: CGFloat
    var height: CGFloat

    @GestureState private var isLongPressing = false

    var body: some View {
        let displayLabel = isLongPressing && keyDef.type == .power ? "√" : keyDef.label

        Button {
            if onLongPress == nil {
                #if canImport(UIKit)
                UIImpactFeedbackGenerator(style: .light).impactOccurred()
                #endif
                onTap()
            }
        } label: {
            Text(displayLabel)
                .font(.title3)
                .fontWeight(.medium)
                .foregroundStyle(textColor)
                .frame(width: width, height: height)
                .background(backgroundColor)
                .clipShape(RoundedRectangle(cornerRadius: 8))
        }
        .buttonStyle(.plain)
        .simultaneousGesture(
            onLongPress != nil ?
            LongPressGesture(minimumDuration: 0.4)
                .updating($isLongPressing) { currentState, gestureState, _ in
                    gestureState = true
                }
                .onEnded { _ in
                    #if canImport(UIKit)
                    UIImpactFeedbackGenerator(style: .medium).impactOccurred()
                    #endif
                    onLongPress?()
                }
            : nil
        )
        .highPriorityGesture(
            onLongPress != nil ?
            TapGesture()
                .onEnded {
                    #if canImport(UIKit)
                    UIImpactFeedbackGenerator(style: .light).impactOccurred()
                    #endif
                    onTap()
                }
            : nil
        )
        .accessibilityLabel(accessibilityLabel)
    }

    private var backgroundColor: Color {
        switch keyDef.type {
        case .digit:
            #if canImport(UIKit)
            return Color(.systemBackground)
            #else
            return Color(NSColor.controlBackgroundColor)
            #endif
        case .operator_, .power, .specialDollar, .specialHash:
            #if canImport(UIKit)
            return Color(.secondarySystemBackground)
            #else
            return Color(NSColor.controlColor)
            #endif
        case .backspace, .enter:
            return Color.red.opacity(0.15)
        }
    }

    private var textColor: Color {
        switch keyDef.type {
        case .digit:
            return .primary
        case .operator_, .power, .specialDollar, .specialHash:
            return .accentColor
        case .backspace, .enter:
            return .red
        }
    }

    private var accessibilityLabel: String {
        switch keyDef.label {
        case "+": return "Add"
        case "−": return "Subtract"
        case "×": return "Multiply"
        case "÷": return "Divide"
        case "^": return "Power, long press for square root"
        case "%": return "Percent"
        case "=": return "Equals"
        case "(": return "Open parenthesis"
        case ")": return "Close parenthesis"
        case ".": return "Decimal point"
        case "⌫": return "Backspace, long press to clear line"
        case "⏎": return "Enter"
        case "$": return "Dollar, long press for variable picker"
        case "#": return "Hash for comment"
        default: return keyDef.label
        }
    }
}
